File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.js
const regex = /\d+\D+/g;

function solution(dartResult) {
    
    const stages = dartResult.match(regex);
    
    const points = [];
    
    for(let i = 0; i < stages.length; i++){
        const stage = stages[i];
        
        const regexNum = /\d+/
        
        let point = stage.match(regexNum)[0];
    
        const rest = stage.replace(regexNum, "");
        
        for(let j = 0; j < rest.length; j++){
            const char = rest[j];
            
            // 숫자 이면 무시 
            if((/[0-9]/).test(char)){
                point = Number(char);
            } else if(char === "S"){
                point = Math.pow(point, 1)
            } else if (char === "D"){
                point = Math.pow(point, 2)
            } else if (char === "T"){
                point = Math.pow(point, 3)
            } else if (char === "*"){
                point *= 2;
                if(points[i - 1]){
                    points[i - 1] *= 2;
                }
                
            } else if (char === "#"){
                point *= (-1)
            }
        }
        points.push(point);
    }
    
    
    return points.reduce((acc, cur) => acc + cur, 0)
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    print(dartResult)
    
    points =[0, 0, 0]
    
    curIdx = 0
    for i in range(len(dartResult)):
        curChar = dartResult[i]
        print("현재 문자", curChar)
        if curChar.isdecimal():
            if i > 0 and not dartResult[i-1].isdecimal():
                curIdx += 1
            
            points[curIdx] = points[curIdx] * 10 + int(curChar)
        else:
            if curChar == "S":
                points[curIdx] = points[curIdx] ** 1
            elif curChar == "D":
                points[curIdx] = points[curIdx] ** 2
            elif curChar == "T":
                points[curIdx] = points[curIdx] ** 3
            elif curChar == "*":
                if curIdx > 0:
                    points[curIdx - 1] = points[curIdx -1] * 2
                points[curIdx] = points[curIdx] * 2
            elif curChar == "#":
                points[curIdx] = points[curIdx] * -1
    return sum(points)